llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/BlacklistingLlapMetricsListener.java

```diff
@@ -0,0 +1,200 @@
+      long maxExecutors = metrics.get(LlapDaemonExecutorInfo.ExecutorNumExecutorsPerInstance.name());
+
+      LOG.debug("Checking node {} with data: " +
+                    "requestHandled={}, " +
+                    "averageTime={}, " +
+                    "emptyExecutors={}, " +
+                    "maxExecutors={}",
+          workerIdentity, requestHandled, averageTime, emptyExecutor, maxExecutors);
+
+      if (maxExecutors == 0) {
+        blacklistedNodes++;
+        if (blacklistedNodes >= this.maxBlacklistedNodes) {
+          LOG.debug("Already too many blacklisted nodes. Skipping.");
+          return;
+        }
+      }
+
+      if (requestHandled > this.minServedTasksNumber) {
+        workerNum++;
+        sumAverageTime += averageTime;
+        if ( averageTime > maxAverageTime) {
+          maxAverageTime = averageTime;
+          maxAverageTimeEmptyExecutors = emptyExecutor;
+          maxAverageTimeMaxExecutors = maxExecutors;
+          maxAverageTimeIdentity = workerIdentity;
+        }
+        sumEmptyExecutors += emptyExecutor;
+      }
+    }
+
+    // If we do not have enough data then return.
+    if (workerNum < 2) {
+      return;
+    }
+
+    LOG.debug("Found slowest node {} with data: " +
+                  "sumAverageTime={}, " +
+                  "sumEmptyExecutors={}, " +
+                  "maxAverageTime={}, " +
+                  "maxAverageTimeEmptyExecutors={}, " +
+                  "maxAverageTimeMaxExecutors={}, " +
+                  "workerNum={}, " +
+                  "maxAverageTimeIdentity={}, " +
+                  "blacklistedNodes={}",
+        sumAverageTime, sumEmptyExecutors, maxAverageTime, maxAverageTimeEmptyExecutors,
+        maxAverageTimeMaxExecutors, workerNum, maxAverageTimeIdentity, blacklistedNodes);
+    // Check if the slowest node is at least timeThreshold times slower than the average
+    long averageTimeWithoutSlowest = (sumAverageTime - maxAverageTime) / (workerNum - 1);
+    if (averageTimeWithoutSlowest * this.timeThreshold < maxAverageTime) {
+      // We have a candidate, let's see if we have enough empty executors.
+      long emptyExecutorsWithoutSlowest = sumEmptyExecutors - maxAverageTimeEmptyExecutors;
+      if (emptyExecutorsWithoutSlowest > maxAverageTimeMaxExecutors * this.emptyExecutorsThreshold) {
+        // Seems like a good candidate, let's try to blacklist
+        try {
+          LOG.debug("Trying to blacklist node: " + maxAverageTimeIdentity);
+          setCapacity(maxAverageTimeIdentity, 0, 0);
+        } catch (Throwable t) {
+          LOG.debug("Can not blacklist node: " + maxAverageTimeIdentity, t);
+        }
+      }
+    }
+  }
+
+  protected void setCapacity(String workerIdentity, int newExecutorNum, int newWaitQueueSize)
+      throws IOException, ServiceException {
+    long currentTime = System.currentTimeMillis();
+    if (currentTime > nextCheckTime) {
+      LlapZookeeperRegistryImpl.ConfigChangeLockResult lockResult =
```

Comment:
Here is the code that opens a window for more blacklisted nodes than the maximum. If multiple AM come to the conclusion to blacklist (different) nodes at the same time, they all might still have -1L in nextCheckTime and therefore go on, trying to call lockForConfigChange. If they are executing in the right order (increasing time), the lock will be successful because it only verifies that the next value is higher than the previous and we will blacklist potentially more daemons than allowed.
Instead, we need to
1) fetch the current value of the distributed long.
2) compare our current timestamp against that and skip if it is smaller.
A local, cached nextCheckTime is unnecessary and can lead to false positives....

Reply:
Good catch! Thanks!